python/hail/linalg/blockmatrix.py

```diff
@@ -553,103 +606,224 @@ def __neg__(self):
         op = getattr(self._jbm, "unary_$minus")
         return BlockMatrix(op())
 
-    @typecheck_method(b=oneof(numeric, block_matrix_type))
+    def _check_common_dim(self, b, op):
+        if self.shape[0] != b.shape[0] and self.shape[1] != b.shape[1]:
+            raise ValueError(f'incompatible shapes for {op}: {self.shape} and {b.shape}')
+
+    # promotes and possibly swaps a and b, such that a is a BlockMatrix
+    # and b is a BlockMatrix, float, or java Array[Double]
+    def _promote(self, b, op):
+        a = self
+        form_a = Form.of(a.shape)
+        reverse = False
+        if isinstance(b, int) or isinstance(b, float):
+            b, form_b = float(b), Form.SCALAR
+        elif isinstance(b, np.ndarray):
+            b = _ndarray_as_2d_float64(b)
+            form_b = Form.of(b.shape)
+            if form_b == Form.SCALAR:
+                b = float(b[0][0])
+            elif form_a == Form.MATRIX and form_b in {Form.COLUMN, Form.ROW}:
```

Comment:
OK, I think we're 98% of the way there.  The user code (e.g. __add__) now looks ideal and your invariants are clear and explicit in the code.

I think you can tighten this up.  In particular, I think you should get to the form for a and b ASAP, and then call yourself recursively if a_form < b_form.  I argued in a previous comment that you should make a function that returns the form from a general value (scalar, numpy array or block matrix).  You can make reverse=False can be a default argument.  (Or not recursive, and do a swap block as you do below).  I think once you do that, the way you want to represent b is determined by its form, in which case you might want to break that out as a helper function.